### PR TITLE
Fix Syntax error with array access in string interpolation

### LINKED ISSUE #62: Syntax error with array access in string interpolation
I have only used this library through another library, but here's a reproducible code:

```
$ composer require opis/closure
```

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

use Opis\Closure\SerializableClosure;

$object = new SerializableClosure(function (): string {
    $a = new \ArrayObject(['foo' => 'bar']);

    return "$a[foo]";
});

$function = unserialize(serialize($object));

var_dump($function());
```

Result (version 3.5.5):

```
PHP Parse error:  syntax error, unexpected '' (T_ENCAPSED_AND_WHITESPACE), expecting '-' or identifier (T_STRING) or variable (T_VARIABLE) or number (T_NUM_STRING) in closure://function (): string {
    $a = new \ArrayObject(['foo' => 'bar']);

    return "$a[\foo]";
} on line 5
```

Result (commit https://github.com/opis/closure/commit/82cc21fefe5f4a56c7c29233129959cb005eb205):

```
string(3) "bar"
```

I think the problematic commit is https://github.com/opis/closure/commit/c17ece00ae2f48bbf640482ed699120af330a31c

Of course a workaround is to "fix" all interpolations like this with

```php
return "{$a['foo']}";
```

### PR DESCRIPTION
**Pull request by abdrzakoxa**
Fix #62

### CODE REVIEW

File: src/ReflectionClosure.php [hunk 1]
```diff
@@ -349,13 +361,17 @@ public function getCode()
...
                         case T_NS_SEPARATOR:
                         case T_STRING:
-                            $id_start = $token[1];
-                            $id_start_ci = strtolower($id_start);
-                            $id_name = '';
-                            $context = 'root';
-                            $state = 'id_name';
-                            $lastState = 'closure';
-                            break 2;
+                            if ($context !== 'string') {
```
Comment (sorinsarca) [inline src/ReflectionClosure.php:364]: This is partially correct. But when we are inside a string we can also have `T_DOLLAR_OPEN_CURLY_BRACES`
```php
namespace Test;
const foo = 0;

$x = [0 => 'Constant foo', 'foo' => 'String foo'];

echo "$x[foo]\n"; // here foo is a string
echo "${x[foo]}\n"; // here foo is a constant
```

Comment (abdrzakoxa) [reply src/ReflectionClosure.php:364]: Yeah where is the problem, I just test it and its work without any problem
I think the problem was here: https://github.com/opis/closure/commit/85fec0b113e163c343c9d643744bc1221641f0b9

Comment (sorinsarca) [reply src/ReflectionClosure.php:364]: It doesn't work

```php
namespace Test;
const foo = 1;

$f = function ($x) {
  return "${x[foo]}";
};

echo \Opis\Closure\serialize($f); // body should contain: return "${x[\Test\foo]}";

```

Comment (abdrzakoxa) [reply src/ReflectionClosure.php:364]: before this commit: https://github.com/opis/closure/commit/85fec0b113e163c343c9d643744bc1221641f0b9
```php
$a = [0 => "bar"];
$c = function() use($a)
{
    return "${a[test_array_key_for_bar_value]}";
};
```
Test result:
> Failed asserting that two strings are equal.
> Expected :'{a[test_array_key_for_bar_value]}'
> Actual   :'bar'
- after this commit: https://github.com/opis/closure/commit/85fec0b113e163c343c9d643744bc1221641f0b9
> Test result:
> OK (1 test, 1 assertion)

Comment (abdrzakoxa) [reply src/ReflectionClosure.php:364]: Hmmm, so body should contain: return "${x[\Test\foo]}"
I will check now

Comment (sorinsarca) [reply src/ReflectionClosure.php:364]: @abdrzakoxa Please read my example. You should notice some things:

a `namespace`, a `const` declaration and NO `define()` used

```php
namespace Test;
const foo = 1;
```

From another file you can `use` the const

```php
use const Test\foo as aliased_foo;
```

File: src/ReflectionClosure.php [hunk 2]
```diff
@@ -306,6 +307,17 @@ public function getCode()
                             }
                             $code .= $token[0];
                             break;
+                        case '"':
+                        case '\'':
```
Comment (sorinsarca) [inline src/ReflectionClosure.php:311]: I don't think that the single quote case should be added because it cannot contain interpolations.
What about heredoc? A test should be probably added.

File: src/ReflectionClosure.php (hunk 1 above)
Comment (abdrzakoxa) [reply src/ReflectionClosure.php:367]: @sorinsarca Solved now

File: src/ReflectionClosure.php [hunk 3]
```diff
@@ -271,6 +272,9 @@ public function getCode()
                         case T_CURLY_OPEN:
                         case T_DOLLAR_OPEN_CURLY_BRACES:
                         case '{':
+                            if ($context === 'string') {
```
Comment (sorinsarca) [inline src/ReflectionClosure.php:275]: This should be only used for T_DOLLAR_OPEN_CURLY_BRACES case.
Anyway, when is the 'string' state "resumed"?

Comment (abdrzakoxa) [reply src/ReflectionClosure.php:275]: I think also should be use in T_CURLY_OPEN
like: `return "{$a[foo]} is $a[0]";`

Comment (sorinsarca) [reply src/ReflectionClosure.php:275]: Probably

**Comment by sorinsarca** [comment]
A complex test

```php
namespace Test;
const foo = 'foo';

$closure = function () {
    $x = [
        'fooConst' => 1,
        'foo' => 3,
    ];
    $y = [
        'foo' => 'Const'
    ];
    return "${x[foo]}${x[foo . "${y[foo]}"]}";
};

echo $closure(), PHP_EOL; //> 31

echo \Opis\Closure\serialize($closure), PHP_EOL; 
//  return "${x[\Test\foo]}${x[\Test\foo . "${y[\Test\foo]}"]}";
```

**Comment by msarca** [comment]
My two cents on this issue: the heuristic nature of the parser should be preserved. It served us well for years and we all knew too well that there are edge cases that won't work. Our purpose was never to build a full PHP parser, just a good enough approximation of it.

**Comment by abdrzakoxa** [comment]
> A complex test
> 
> ```
> namespace Test;
> const foo = 'foo';
> 
> $closure = function () {
>     $x = [
>         'fooConst' => 1,
>         'foo' => 3,
>     ];
>     $y = [
>         'foo' => 'Const'
>     ];
>     return "${x[foo]}${x[foo . "${y[foo]}"]}";
> };
> 
> echo $closure(), PHP_EOL; //> 31
> 
> echo \Opis\Closure\serialize($closure), PHP_EOL; 
> //  return "${x[\Test\foo]}${x[\Test\foo . "${y[\Test\foo]}"]}";
> ```

I just test this and it's passed
```php
    const bar = 'bar';
    public function testComplexConstant()
    {
        $c = function () {
            $x = [
                'barConst' => 1,
                'bar' => 3,
            ];
            $y = [
                'bar' => 'Const'
            ];
            return "${x[bar]}${x[bar . "${y[bar]}"]}";
        };
        $u = $this->s($c);
        self::assertEquals($c(), $u());
    }
```

**Comment by sorinsarca** [comment]
> I just test this and it's passed

My bad, I meant:

```php
namespace Test;

const foo = 'fooVal';

$closure = function () {
    $x = [
        'fooValConst' => 1,
        'fooVal' => 3,
        'foo' => 4,
    ];
    $y = [
        'fooVal' => 'Const'
    ];
    return "${x[foo]}${x[foo . "${y[foo]}"]}$x[foo]${x[foo]}";
};

echo $closure(), PHP_EOL; //> 3143

echo \Opis\Closure\serialize($closure), PHP_EOL;
// return "{x[\Test\foo]}{x[\Test\foo . "{y[\Test\foo]}"]}$x[foo]{x[\Test\foo]}";

```

Please test the generated source code, not the results of the function call.

**Comment by abdrzakoxa** [comment]
> > I just test this and it's passed
> 
> My bad, I meant:
> 
> ```
> namespace Test;
> 
> const foo = 'fooVal';
> 
> $closure = function () {
>     $x = [
>         'fooValConst' => 1,
>         'fooVal' => 3,
>         'foo' => 4,
>     ];
>     $y = [
>         'fooVal' => 'Const'
>     ];
>     return "${x[foo]}${x[foo . "${y[foo]}"]}$x[foo]${x[foo]}";
> };
> 
> echo $closure(), PHP_EOL; //> 3143
> 
> echo \Opis\Closure\serialize($closure), PHP_EOL;
> // return "{x[\Test\foo]}{x[\Test\foo . "{y[\Test\foo]}"]}$x[foo]{x[\Test\foo]}";
> ```
> 
> Please test the generated source code, not the results of the function call.

But also this was passed
```php
       const foo = 'fooVal';
       $c = function () {
            $x = [
                'fooValConst' => 1,
                'fooVal' => 3,
                'foo' => 4,
            ];
            $y = [
                'fooVal' => 'Const'
            ];
            return "${x[foo]}${x[foo . "${y[foo]}"]}$x[foo]${x[foo]}";
        };
        $u = $this->s($c);
        echo $u(); // 3143
```
> pls make sure you have latest changes for this branch

File: src/ReflectionClosure.php (hunk 2 above)
Comment (abdrzakoxa) [reply src/ReflectionClosure.php:314]: Right, I will work for this

**Comment by abdrzakoxa** [comment]
@sorinsarca 
All possible problems now are solved for this issue
You can review it now

File: src/ReflectionClosure.php [hunk 4]
```diff
@@ -271,6 +272,9 @@ public function getCode()
                         case T_CURLY_OPEN:
                         case T_DOLLAR_OPEN_CURLY_BRACES:
                         case '{':
+                            if ($context === 'string' && $token[0] !== '{') {
```
Comment (sorinsarca) [inline src/ReflectionClosure.php:275]: This context switching should happen only when T_DOLLAR_OPEN_CURLY_BRACES.
Something like this:
```php
case T_DOLLAR_OPEN_CURLY_BRACES:
  if ($context === 'string') {
    $context = 'root';
  }
 // no break
case '{':
  $code .= ...
```

File: src/ReflectionClosure.php [hunk 5]
```diff
@@ -306,6 +310,21 @@ public function getCode()
                             }
                             $code .= $token[0];
                             break;
+                        case T_START_HEREDOC:
```
Comment (sorinsarca) [inline src/ReflectionClosure.php:313]: This should be two distinct cases.

File: src/ReflectionClosure.php (hunk 4 above)
Comment (abdrzakoxa) [reply src/ReflectionClosure.php]: Yes, but also T_CURLY_OPEN because:
- after remove: T_CURLY_OPEN
```php
$c = function () {
    $user = ['username' => 'abdzakoxa'];
    return "{$user['username']}";
};
$u = $this->s($c);
```
- test result:
ParseError : syntax error, unexpected end of file
 closure://function () {
             $user = ['username' => 'abdzakoxa'];
             return "{$user['username']}:4

File: src/ReflectionClosure.php (hunk 5 above)
Comment (abdrzakoxa) [reply src/ReflectionClosure.php:313]: Yeah, I did
